upload/upload.go

```diff
@@ -0,0 +1,201 @@
+	ContentType string
+
+	// URL to upload the parts to
+	URL string
+
+	// Job information about the upload which will be saved into a file named job.json in the same folder
+	Job map[string]interface{}
+}
+
+type part struct {
+	index  int
+	reader io.Reader
+}
+
+func newClient() (httpclient.Client, error) {
+	hcConfig := &httpclient.Config{
+		Paginator: httpclient.NoPaginator(),
+		Retryable: httpclient.NewBackoffRetry(10*time.Millisecond, 100*time.Millisecond, 30*time.Second, 2.0),
+	}
+	return api.NewHTTPAPIClient(hcConfig)
+}
+
+func upload(opts Options, urlpath string, reader io.Reader) error {
+	client, err := newClient()
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest(http.MethodPut, urlpath, reader)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Cookie", strings.Join(opts.Cookies, "; "))
+	req.Header.Set("x-amz-acl", "bucket-owner-full-control")
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode == http.StatusOK {
+		io.Copy(ioutil.Discard, resp.Body)
+		resp.Body.Close()
+		return nil
+	}
+	return fmt.Errorf("error uploading %v. status code was %v after multi attempts", urlpath, resp.StatusCode)
+}
+
+// Upload a file to the upload server in multi part upload
+func Upload(opts Options) (int, error) {
+	if opts.PartSize <= 0 || opts.PartSize < MinUploadPartSize {
+		opts.PartSize = DefaultUploadPartSize
+	}
+	if opts.Concurrency <= 0 {
+		opts.Concurrency = DefaultUploadConcurrency
+	}
+	if len(opts.Cookies) == 0 {
+		return 0, fmt.Errorf("missing required Cookies")
+	}
+	if opts.AttemptsForPart <= 0 || opts.AttemptsForPart > MaxAttemptsForPart {
+		opts.AttemptsForPart = MaxAttemptsForPart
+	}
+	if opts.Body == nil {
+		return 0, fmt.Errorf("missing required Body")
+	}
+	if opts.ContentType == "" {
+		return 0, fmt.Errorf("missing required ContentType")
+	}
+	if opts.URL == "" {
+		return 0, fmt.Errorf("missing required URL")
+	}
+	var wg sync.WaitGroup
+	ch := make(chan part, opts.Concurrency)
+	errors := make(chan error, opts.Concurrency)
+	for i := 0; i < opts.Concurrency; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for r := range ch {
+				if err := upload(opts, fmt.Sprintf("%s.%d", opts.URL, r.index), r.reader); err != nil {
+					errors <- err
+					return
+				}
+			}
+		}()
+	}
+	var index int
+	// split the buffer in chunks of PartSize and upload each part separately
+	// in its own goroutine
+	for {
+		buf := make([]byte, opts.PartSize)
+		n, err := opts.Body.Read(buf)
+		if err == io.EOF || n < 0 {
+			break
+		}
+		if err != nil {
+			return 0, err
+		}
+		var ok bool
+		for !ok {
+			select {
+			case ch <- part{reader: bytes.NewBuffer(buf[:n]), index: index}:
+				ok = true
+				break
+			default:
+				// check to make sure not in an error state
+				select {
+				case err := <-errors:
+					return 0, err
+				default:
+				}
+				time.Sleep(time.Microsecond)
+			}
+		}
+		index++
+	}
+	close(ch)
+	wg.Wait()
```

Comment:
This looks suspicious, I think it may deadlock if more than 1 upload part returns error. Upload will wait for errors channel to empty, so wg will not be Done, and on this like we wait for wg.

Reply:
i won't because of the code above that checks for error.